clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
         }
     }
 
-}
\ No newline at end of file
+    private void maybeThrowFencedInstanceException() {
+        if (isFenced) {
+            throw new FencedInstanceIdException("Get fenced exception for group.instance.id " +
+                groupInstanceId.orElse("null"));
+        }
+    }
+
+    // Visible for testing
+    void maybeInvokeCallbacks() {
+        if (callbacks() > 0) {
+            invoker.executeCallbacks();
+        }
+    }
+
+    // Visible for testing
+    int callbacks() {
+        return invoker.callbackQueue.size();
+    }
+
+    /**
+     * Utility class that helps the application thread to invoke user registered {@link OffsetCommitCallback}. This is
+     * achieved by having the background thread to register a {@link OffsetCommitCallbackTask} to the invoker upon the
+     * future completion, and execute the callbacks when user polls/commits/closes the consumer.
+     */
+    private class OffsetCommitCallbackInvoker {
+        // Thread-safe queue to store callbacks
+        private final BlockingQueue<OffsetCommitCallbackTask> callbackQueue = new LinkedBlockingQueue<>();
+
+        public void submit(final OffsetCommitCallbackTask callback) {
+            try {
+                callbackQueue.offer(callback);
+            } catch (Exception e) {
+                log.error("Failed to enqueue OffsetCommitCallback", e);
+            }
+        }
+
+        public void executeCallbacks() {
+            LinkedList<OffsetCommitCallbackTask> callbacks = new LinkedList<>();
+            callbackQueue.drainTo(callbacks);
+            while (!callbacks.isEmpty()) {
+                OffsetCommitCallbackTask callback = callbacks.poll();
+                if (callback != null) {
+                    callback.invoke();
+                }
+            }
+        }
+    }
+
+    private class OffsetCommitCallbackTask {
```

Comment:
Should be a static class.

Reply:
Hi, it is not static for the reason of the use of isFenced.  Let me know what do you think.